File: algorithm/c++/SafetyZone_2468/safety_2468_stack.cpp
#include <iostream>
#include <algorithm>
#include <utility>
#include <stack>
#include <cstring>

using namespace std;

int N;
int map[100][100];
bool visited[100][100];
int dr[4] = {0,1,0,-1};
int dc[4] = {1,0,-1,0};
int max_cnt;
int min_h;
int max_h;

void dfs(int r, int c, int h)
{
    stack<pair<int,int>> stack;
    stack.push(make_pair(r,c));
    visited[r][c] = true;
    
    while(!stack.empty())
    {
        auto [cur_r, cur_c] = stack.top();
        stack.pop();
        
        for (int i = 0; i < 4; i++)
        {
            int new_r = cur_r + dr[i];
            int new_c = cur_c + dc[i];

            if (new_r >= 0 && new_c >= 0 && new_r < N && new_c < N && map[new_r][new_c] > h && !visited[new_r][new_c])
            {
                stack.push(make_pair(new_r, new_c));
                visited[new_r][new_c] = true;
            }
        }
    }
}

int main()
{
    cin >> N;

    for (int i = 0; i < N*N; i++)
    {
        int h; cin >> h;
        max_h = max(h,max_h);
        min_h = min(h,min_h);

        map[i/N][i%N] = h;
    }

    for (int h = min_h-1; h < max_h; h++)
    {
        int cnt = 0;

        for (int i = 0; i < N*N; i++)
        {
            int cur_r = i / N;
            int cur_c = i % N;

            if (map[cur_r][cur_c] > h && !visited[cur_r][cur_c])
            {
                dfs(cur_r, cur_c, h);
                cnt++;
            }    
        }
        
        max_cnt = max(cnt, max_cnt);
        memset(visited, false, sizeof(visited));
    }
    
    cout << max_cnt << endl;

    return 0;
}

File: algorithm/java/Java_Algorithm/src/Safety2468/Safety2468stack.java
package Safety2468;

import java.io.*;
import java.util.*;

public class Safety2468stack {
    
    static int N, minH, maxH, maxCnt;
    static int[][] map;
    static boolean[][] visited;
    static int dr[] = {0,1,0,-1};
    static int dc[] = {1,0,-1,0};
    
    static class Pair {
        int row,col;

        public Pair(int row, int col) {
            this.row = row;
            this.col = col;
        }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        N = Integer.parseInt(br.readLine());
        map = new int[N][N];
        visited = new boolean[N][N];

        for (int i = 0; i < N; i++) {
            StringTokenizer stk = new StringTokenizer(br.readLine(), " ");
            int index = 0;

            while(stk.hasMoreTokens()) {
                int curH = Integer.parseInt(stk.nextToken());
                minH = Math.min(minH, curH);
                maxH = Math.max(maxH, curH);

                map[i][index++] = curH;
            }
        }
        
        for (int h = minH-1; h < maxH; h++) {
            
            int cnt = 0;
            
            for (int i = 0; i < N*N; i++) {
                int curR = i / N;
                int curC = i % N;

                if (map[curR][curC] > h && !visited[curR][curC]) {
                    dfs(curR, curC, h);
                    cnt++;
                }
            }

            maxCnt = Math.max(maxCnt, cnt);
            visited = new boolean[N][N];
        }
        
        bw.write(maxCnt + "");
        bw.close();
        br.close();
    }

    static void dfs(int row, int col, int height) {
        
        Stack<Pair> stack = new Stack<>();
        stack.push(new Pair(row, col));
        visited[row][col] = true;

        while(!stack.isEmpty()) {
            int curR = stack.peek().row;
            int curC = stack.peek().col;
            stack.pop();

            for (int i = 0; i < 4; i++) {
                int newR = curR + dr[i];
                int newC = curC + dc[i];

                if (newR >= 0 && newC >= 0 && newR < N && newC < N && !visited[newR][newC] && map[newR][newC] > height) {
                    stack.push(new Pair(newR, newC));
                    visited[newR][newC] = true;
                }
            }
        }
    }
}



